exonum/src/encoding/protobuf/mod.rs

```diff
@@ -0,0 +1,245 @@
+    fn to_pb(&self) -> Self::ProtoStruct;
+
+    /// ProtoStruct -> Struct
+    fn from_pb(pb: Self::ProtoStruct) -> Result<Self, ()>;
+}
+
+impl<T> BinaryForm for T
+where
+    T: ToProtobuf,
+{
+    fn encode(&self) -> Result<Vec<u8>, Error> {
+        Ok(self.to_pb().write_to_bytes().unwrap())
+    }
+
+    fn decode(buffer: &[u8]) -> Result<Self, Error> {
+        let mut pb = <Self as ToProtobuf>::ProtoStruct::new();
+        pb.merge_from_bytes(buffer).unwrap();
+        Self::from_pb(pb).map_err(|_| "Conversion from protobuf error".into())
+    }
+}
+
+/// Used for establishing correspondence between rust struct field
+/// and protobuf value
+pub trait ProtobufValue: Sized {
+    /// Type of value that is returned with pb.take_field() method
+    type ProtoValue;
+    /// RustField -> ProtobufRustField
+    fn to_pb_field(&self) -> Self::ProtoValue;
+    /// ProtobufRustField -> RustField
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()>;
+}
+
+impl<T> ProtobufValue for T
+where
+    T: ToProtobuf,
+{
+    type ProtoValue = <Self as ToProtobuf>::ProtoStruct;
+
+    fn to_pb_field(&self) -> Self::ProtoValue {
+        self.to_pb()
+    }
+
+    fn from_pb_field(pb: Self::ProtoValue) -> Result<Self, ()> {
+        Self::from_pb(pb)
+    }
+}
+
+impl ToProtobuf for crypto::Hash {
+    type ProtoStruct = Hash;
+
+    fn to_pb(&self) -> Hash {
+        let mut hash = Hash::new();
+        hash.set_data(self.as_ref().to_vec());
+        hash
+    }
+
+    fn from_pb(pb: Hash) -> Result<Self, ()> {
+        let data = pb.get_data();
+        crypto::Hash::from_slice(data).ok_or(())
+    }
+}
+
+impl ToProtobuf for crypto::PublicKey {
+    type ProtoStruct = PublicKey;
+
+    fn to_pb(&self) -> PublicKey {
+        let mut key = PublicKey::new();
+        key.set_data(self.as_ref().to_vec());
+        key
+    }
+
+    fn from_pb(pb: PublicKey) -> Result<Self, ()> {
+        let data = pb.get_data();
+        crypto::PublicKey::from_slice(data).ok_or(())
+    }
+}
+
+impl ToProtobuf for bit_vec::BitVec {
+    type ProtoStruct = BitVec;
+
+    fn to_pb(&self) -> BitVec {
+        let mut bit_vec = BitVec::new();
+        bit_vec.set_data(self.to_bytes());
+        bit_vec.set_len(self.len() as u64);
+        bit_vec
+    }
+
+    fn from_pb(pb: BitVec) -> Result<Self, ()> {
+        let data = pb.get_data();
+        let mut bit_vec = bit_vec::BitVec::from_bytes(data);
+        bit_vec.truncate(pb.get_len() as usize);
+        Ok(bit_vec)
+    }
+}
+
+impl ToProtobuf for DateTime<Utc> {
+    type ProtoStruct = well_known_types::Timestamp;
+
+    fn to_pb(&self) -> well_known_types::Timestamp {
+        let mut ts = well_known_types::Timestamp::new();
+        ts.set_seconds(self.timestamp());
+        ts.set_nanos(self.timestamp_subsec_nanos() as i32);
+        ts
+    }
+
+    fn from_pb(pb: well_known_types::Timestamp) -> Result<Self, ()> {
+        Ok(Utc.timestamp(pb.get_seconds(), pb.get_nanos() as u32))
```

Comment:
Should we test that `pb.get_nanos() < 1e9`?

Reply:
I thought that i32 is always big enough for u32(if it is => 0, of course). Or do you mean that we should test it if it is < 0?

Reply:
@slowli is talking about verifying that nanosecond value has correct range [0; 1e9) as expected by `Utc` constructor. It seems to validate the value itself and [will panic with an invalid nanosecond value](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=65a4a51d8b44eca13689db1c7b27be91). I guess we should use `timestamp_opt()` and return an `Err` here.